pkg/webhook/mutation/pod/metadata/workload.go

```diff
@@ -99,19 +78,18 @@ func findRootOwner(ctx context.Context, clt client.Client, partialObjectMetadata
 					"namespace", objectMetadata.Namespace,
 				)
 
-				return newWorkloadInfo(partialObjectMetadata), err
+				return newWorkloadInfo(wellKnownOwner), err
 			}
 
-			return findRootOwner(ctx, clt, ownerObjectMetadata)
-		}
-	}
+			if isWellKnownWorkload(owner) {
+				wellKnownOwner = ownerObjectMetadata
+			}
 
-	if !foundController {
-		// Out of all owner references no one is a Controller
-		return newUnknownWorkloadInfo(), nil
+			return findRootOwner(ctx, clt, ownerObjectMetadata, wellKnownOwner)
+		}
 	}
 
-	return newWorkloadInfo(partialObjectMetadata), nil
+	return newWorkloadInfo(wellKnownOwner), nil
```

Comment:
Here is a draft of the solution I was talking about:
```go
package metadata

import (
	"context"

	dtwebhook "github.com/Dynatrace/dynatrace-operator/pkg/webhook"
	"github.com/pkg/errors"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

type workloadInfo struct {
	name string
	kind string
}

func newWorkloadInfo(partialObjectMetadata *metav1.PartialObjectMetadata) *workloadInfo {
	name := partialObjectMetadata.ObjectMeta.Name
	if name == "" {
		name = partialObjectMetadata.ObjectMeta.GenerateName
	}
	return &workloadInfo{
		name: name,
		kind: partialObjectMetadata.Kind,
	}
}

func (mut *Mutator) retrieveWorkload(request *dtwebhook.MutationRequest) (*workloadInfo, error) {
	workload, err := findRootOwnerOfPod(request.Context, mut.metaClient, request.Pod, request.Namespace.Name)
	if err != nil {
		return nil, err
	}

	return workload, nil
}

func findRootOwnerOfPod(ctx context.Context, clt client.Client, pod *corev1.Pod, namespace string) (*workloadInfo, error) {
	podPartialMetadata := &metav1.PartialObjectMetadata{
		TypeMeta: metav1.TypeMeta{
			APIVersion: pod.APIVersion,
			Kind:       pod.Kind,
		},
		ObjectMeta: metav1.ObjectMeta{
			Name: pod.ObjectMeta.Name,
			// pod.ObjectMeta.Namespace is empty yet
			Namespace:       namespace,
			OwnerReferences: pod.ObjectMeta.OwnerReferences,
		},
	}

	rootOwner, err := findRootOwner(ctx, clt, podPartialMetadata)
	if err != nil {
		return nil, errors.WithStack(err)
	}

	return newWorkloadInfo(rootOwner), nil
}

func findRootOwner(ctx context.Context, clt client.Client, childObjectMetadata *metav1.PartialObjectMetadata) (parentObjectMetadata *metav1.PartialObjectMetadata, err error) {
	objectMetadata := childObjectMetadata.ObjectMeta
	for _, owner := range objectMetadata.OwnerReferences {
		if owner.Controller != nil && *owner.Controller {
			parentObjectMetadata = &metav1.PartialObjectMetadata{
				TypeMeta: metav1.TypeMeta{
					APIVersion: owner.APIVersion,
					Kind:       owner.Kind,
				},
			}

			err = clt.Get(ctx, client.ObjectKey{Name: owner.Name, Namespace: objectMetadata.Namespace}, parentObjectMetadata)
			if err != nil {
				log.Error(err, "failed to query the object",
					"apiVersion", owner.APIVersion,
					"kind", owner.Kind,
					"name", owner.Name,
					"namespace", objectMetadata.Namespace,
				)

				return childObjectMetadata, err
			}

			parentObjectMetadata, err = findRootOwner(ctx, clt, parentObjectMetadata)
			if err != nil {
				return childObjectMetadata, err
			}
			if isWellKnownWorkload(parentObjectMetadata){
				return parentObjectMetadata, nil
			}
			return childObjectMetadata, nil
		}
	}

	return childObjectMetadata, nil
}

func isWellKnownWorkload(owner *metav1.PartialObjectMetadata) bool {
	knownWorkloads := []metav1.TypeMeta{
		{Kind: "ReplicaSet", APIVersion: "apps/v1"},
		{Kind: "Deployment", APIVersion: "apps/v1"},
		{Kind: "ReplicationController", APIVersion: "v1"},
		{Kind: "StatefulSet", APIVersion: "apps/v1"},
		{Kind: "DaemonSet", APIVersion: "apps/v1"},
		{Kind: "Job", APIVersion: "batch/v1"},
		{Kind: "CronJob", APIVersion: "batch/v1"},
		{Kind: "DeploymentConfig", APIVersion: "apps.openshift.io/v1"},
	}

	for _, knownController := range knownWorkloads {
		if owner.Kind == knownController.Kind &&
			owner.APIVersion == knownController.APIVersion {
			return true
		}
	}

	return false
}
```

sidenote: we talked about it yesterday, I felt there was a better way to do this, so I quickly came up with this, (didn't test it)

Reply:
applied